lessons-02 update (#91)
mimuw-jnp2-rust/mimuw-jnp2-rust.github.io: content/lessons/02_ownership/index.md

Let's expand on the last example. The scoping is not really important for that one, so we don't include it here.

```rust
// Same thing, 's' is now an owner.
let s = String::from("a string");

// Easy, 's2' becomes another owner... right?
let s2 = s;

// This doesn't work, can you guess why?
println!("And the contents are: {}", s);
```

> At first glance everything looks great. If we write this code (well, an equivalent of it) in basically any other popular language, it will compile no issue - but it does *not* compile here and there's a good reason why.

Reviewer: ```suggestion
At first glance everything looks great. If we write this code (well, an equivalent of it) in basically any other popular language, it will compile without issues - but it does *not* compile here and there's a good reason why.
```

}
```

> Thankfully, Rust allows us to access a resource without becoming an owner through the use of references and the `&` operator. This is called a borrow.

Reviewer: Why remove the real-life example? I'd keep it.

### Exercise

> Our previous solution using `clone()` was pretty inefficient. How should you modify this this code now?

Reviewer: ```suggestion
Our previous solution using `clone()` was pretty inefficient. How should you modify this code now?
```

  count_animals(1, s.clone());
  count_animals(2, s.clone());

>   count_animals(3, s); // We previously could've ommitted the clone() here. Why?

Reviewer: ```suggestion
  count_animals(3, s); // We previously could've omitted the clone() here. Why?
```

  count_animals(1, &s);
  count_animals(2, &s);
  count_animals(3, &s);
  count_animals(4, "goat"); // In this version of the code it doesn't compile.
}
```

### Further reading

- [The Book, chapter 4](https://doc.rust-lang.org/stable/book/ch04-00-understanding-ownership.html) - obligatory reading

> - [Rust by Example, chapter Scoping rules](https://doc.rust-lang.org/stable/rust-by-example/index.html) - obligatory reading

Reviewer: ♻️ Instead of annotating each link with `obligatory/additional reading`, let's have two subsections.
